### PR TITLE
[WIP] Gradient boosting

### PR DESCRIPTION
**Pull request by ogrisel**
WIP experimental class to implement l2 loss GBM and benchmark it on the Higgs boson dataset.

There are many todos and it lacks test but I open this PR to make this work more visible.

For this small benchmark, pygbm is approximately 2x slower than LightGBM (serial learner) on a laptop with 2 physical cores (4 hyperthreads).

I lack a good numba profiler (see https://github.com/numba/numba/issues/3254) to understand why it's slower than LightGBM.

Note that the final AUC is not as close but slightly lower than LightGBM with what I thought were the same hyperparameters (31 leaf nodes, 10 trees, min_hessian_split=1e-3, min_data_leaf=20). There is probably something wrong in my code that does not behave as LightGBM's algo. We would need to introspect the actual splits to check where the two implementations differ.

### CODE REVIEW

**Comment by ogrisel** [comment]
The tests need scikit-learn master to pass.

**Comment by ogrisel** [comment]
If people are curious, here is the output of a benchmark run:

```
Training set with 10950000 records with 28 features.
Fitting a LightGBM model...
[LightGBM] [Info] Total Bins 6143
[LightGBM] [Info] Number of data: 10950000, number of used features: 28
[LightGBM] [Info] Start training from score 0.529920
[LightGBM] [Debug] Trained a tree with leaves = 31 and max_depth = 7
[LightGBM] [Debug] Trained a tree with leaves = 31 and max_depth = 7
[LightGBM] [Debug] Trained a tree with leaves = 31 and max_depth = 8
[LightGBM] [Debug] Trained a tree with leaves = 31 and max_depth = 7
[LightGBM] [Debug] Trained a tree with leaves = 31 and max_depth = 9
[LightGBM] [Debug] Trained a tree with leaves = 31 and max_depth = 7
[LightGBM] [Debug] Trained a tree with leaves = 31 and max_depth = 9
[LightGBM] [Debug] Trained a tree with leaves = 31 and max_depth = 8
[LightGBM] [Debug] Trained a tree with leaves = 31 and max_depth = 9
[LightGBM] [Debug] Trained a tree with leaves = 31 and max_depth = 7
done in 23.412s, ROC AUC: 0.7642078442392614
JIT compiling code for the pygbm model...
done in 3.541s
Fitting a pygbm model...
Binning 1.226 GB of data: 8.545 s (143.522 MB/s)
Fitting gradient boosted rounds:
[0/10]
[1/10], 2.830s/tree
[2/10], 3.076s/tree
[3/10], 3.075s/tree
[4/10], 3.042s/tree
[5/10], 3.001s/tree
[6/10], 2.994s/tree
[7/10], 2.968s/tree
[8/10], 3.004s/tree
[9/10], 2.992s/tree
Fit 10 trees in 39.050 s, (310 total leaf nodes)
done in 39.097s, ROC AUC: 0.7594941477318347
```

**Comment by ogrisel** [comment]
This still lacks some test but let's merge so that others can issue pull requests to improve upon that base.
